Sources/Private/CoreAnimation/CoreAnimationLayer.swift

```diff
@@ -441,6 +441,11 @@ extension CoreAnimationLayer: RootAnimationLayer {
     // Unimplemented / unused
   }
 
+  func getHierarchyKeypaths() -> [String] {
+    logHierarchyKeypaths()
+    return []
```

Comment:
Can you implement this for the Core Animation rendering engine as well?

Reply:
Yes, will look into that.

Reply:
I see how displayIfNeeded() inside logHierarchyKeypaths in CoreAnimationLayer is magically causing the log to happen. Looks to me that customKeyframes is somehow being triggered. It doesn't seem to be straightforward to implement getHierarchyKeypaths for Core Animation rendering.

Reply:
Here's an approach we use elsewhere to get an array of keypath names just by calling `logHierarchyKeypaths()`: https://github.com/airbnb/lottie-ios/blob/master/Tests/AnimationKeypathTests.swift#L118 This works on both the Core Animation and Main Thread rendering engines.

Reply:
Thanks, looked into it and I thought it would be best to create an extension to RootAnimationLayer to have the same implementation for the function and use the idea from hierarchyKeypaths from the code in your link.
This implementation causes a crash when returning in getHierarchyKeypaths.

added to protocol RootAnimationLayer:
`func getHierarchyKeypaths(getAnimationView: @escaping (_ newLogger: LottieLogger) -> LottieAnimationView) -> [String]
`

created extension in RootAnimationLayer.swift to have default implementation for the function:

```
extension RootAnimationLayer {
    func getHierarchyKeypaths(getAnimationView: @escaping (_ newLogger: LottieLogger) -> LottieAnimationView) -> [String] {
        var printedMessages = [String]()
        let logger = LottieLogger(info: { message in
            printedMessages.append(message())
        })
       let animationView = getAnimationView(logger)
       animationView.logHierarchyKeypaths()
       
      return Array(printedMessages[1...])    //Thread 1: Fatal error: Range requires lowerBound <= upperBound
    }
}

```

In LottieAnimationView.swift:

```
private func getAnimationView(newLogger: LottieLogger) -> LottieAnimationView {
    self.logger = newLogger
    return self
}
    
public var allAnimationKeypaths: [String] {
    animationLayer?.getHierarchyKeypaths(getAnimationView: getAnimationView) ?? []
}
```

Reply:
If you push a commit with that change I can take a look. It sounds like the problem is that `printedMessages` is empty.

Reply:
It's possible that `logHierarchyKeypaths()` is asynchronous, so hasn't finished running by the time you try to read `printedMessages`.

Reply:
Yes that's right, the list is empty. I have pushed the changes. Everything is inside allAnimationKeypaths in LottieAnimationView. Don't see that logHierarchyKeypaths is async.

Reply:
The newest commit solves everything, both for core and main thread engines. Really appreciate the support and feedback. Just let me know if something else needs to be changed.